languages/bash.scm

```diff
@@ -22,23 +24,43 @@
   "fi"
   (string)
   (test_command)
+  ; TODO: etc.
 ] @prepend_space
 
 ;; Commands
 
-; One command per line, modulo many exceptions:
-; * Strings of && and ||
-; * Pipelines
-; * Subshells
-; * As arguments to if, while, for, etc.
-; * Others?...
+; NOTE "Command" is shorthand for a "unit of execution":
+; * Simple commands (e.g., binaries, builtins, functions, etc.)
+; * Command lists
+; * Command pipelines
 ;
-; FIXME Can this be generalised, or does *every* context need to be
-; individually enumerated?...
-(program (command) @prepend_spaced_softline @append_hardline)
-(if_statement _ "then" (command) @prepend_spaced_softline)
-(elif_clause _ "then" (command) @prepend_spaced_softline)
-(else_clause (command) @prepend_spaced_softline)
+; That is: [(command) (list) (pipeline)], per the grammar
+
+; FIXME I don't think it's possible to insert the necessary line
```

Comment:
Should `command` be treated as a leaf (so we leave it entirely alone)?

Reply:
I'm not fully clear about the consequences of `@leaf`. I understand that Topiary just skips over these nodes, but to what end/what effect would that have on `(command)`?

Reply:
It would keep leafs formatted as is. I would think that might be better than forcing all commands to be single-line.

Reply:
Ah, I see...

So the following:

```bash
foo \
--bar
```

Will be formatted as `foo --bar`, without making it a leaf. Whereas, as a leaf, like you say, no formatting changes are applied. That's certainly better, assuming that's what the author intended :+1:

Changed in cfbd8cb2cd64568

Reply:
...Oh, but then that obviates this rule:

```scheme
; Space between command line arguments
(command
  argument: _* @append_space @prepend_space
)
```

Which formats, e.g., `foo     --bar  --quux` into `foo --bar --quux`... I quite like this rule.

Reply:
Yeah, I see.